site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+
+![decorator pattern](https://upload.wikimedia.org/wikipedia/commons/8/83/W3sDesign_Decorator_Design_Pattern_UML.jpg?1591628685344)
+
+In simpler terms, let's say we have an object called `Cow` that has some methods:
+
+![cow undecorated](/blog/2020-06-10-opentracing/cowundecorated.png) 
+
+We want to introduce additional functionality on top of what `Cow` already does, without modifying the actual code of the `Cow` object. For example, we want to measure performance of each method, and log the parameters that are being passed to each method. Here's how it would look if we apply the [decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern):
+
+![cow decorated](/blog/2020-06-10-opentracing/cowdecorated.png) 
+
+We wrapped each method of `Cow` in a chain of additional functions `f(x) = y` became `f(x) = a(b(y))`, each function having it's own responsibility.
+
+If we apply the same pattern to our problem, we can decorate all of `mattermost-server` API calls with OpenTracing, without actually modifying the functions themselves!
+
+Implementing such functionality in other, dynamic languages is rather trivial. For example, here's how JavaScript handles it. Given a simple `Cow` object:
+
+```javascript
+const cow = {
+  feed: function(x) {
+    return `Ate for ${x} seconds!`
+  },
+  speak: function(x) {
+    return `${"Moo ".repeat(x)}!`
+  }
+}
+
+console.log(cow.feed(20))
+console.log(cow.speak(3))
+```
+
+We can wrap it in a [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy):
+
+```javascript
+const tracerHandler = {
+  get: function(target, prop, receiver) {
+    if (typeof target[prop] === "function") {
+      return function(...args) {
+        console.log(`'${prop} 'called with arguments: `, ...arguments);
+        return target[prop](...arguments);
+      };
+    }
+  }
+};
+
+const timerHandler = {
+  get: function(target, prop, receiver) {
+    if (typeof target[prop] === "function") {
+      return function(...args) {
+        console.log(`starting '${prop}'`);
+        const t1 = window.performance.now();
+
+        const res = target[prop](...arguments);
+        const t2 = window.performance.now();
+
+        console.log(`'${prop}' took ${t2 - t1}ns`);
+        return res;
+      };
+    }
+  }
+};
+
+const proxy = new Proxy(cow, tracerHandler);
+const proxy2 = new Proxy(proxy, timerHandler);
+console.log(proxy2.feed(20));
+console.log(proxy2.speak(3));
+```
+
+Unfortunately, in Go, there's no way to do this in a performant manner (the regular approach would involve using reflection and that can have a serious performance impact on the underlying code)
+
+## Our solution
+
+Implementation of decorator pattern that we chose involved 3 parts:
+
+1. Struct embedding
+2. Code parsing using AST
+3. Code generation using templates
+
+### Struct embedding
+
+Since Go is not object-oriented, it uses Struct embedding as a type of inheritance. Here's a small example:
```

Comment:
No need to capitalize "Struct".

Also, might be clearer if you could expand of what is meant by "type of inheritance". Does this mean there are various types of inheritances ? Or embedding is a loose association of inheritance?